src/main/multiparser.js

```diff
@@ -20,8 +20,6 @@ async function printEmbeddedLanguages(
     (printer.isNode ?? printer.canAttachComment)?.bind(printer) ?? (() => true);
   const hasPrettierIgnore =
     printer.hasPrettierIgnore?.bind(printer) ?? (() => false);
-  const detectEmbeddedLanguage =
-    printer.detectEmbeddedLanguage?.bind(printer) ?? (() => true);
   const { ignoredProperties } = printer.massageAstNode;
 
   const pathStacks = [];
```

Comment:
This variable needs a better name. `embedCallResults`?

Reply:
Yah, need more work, just testing if it works.

Reply:
It looks good. One more name idea: `embedThunks`

Reply:
How about this https://github.com/thorn0/prettier/pull/1/commits/5c1a383646f69a3d83cf32cb32cedf9be014765a 😄

Reply:
Doesn't look like it might work :)

Reply:
And this https://github.com/thorn0/prettier/pull/1/commits/8ae161b98b82a200a97b19cb11b49774ed626a2c 😄

Reply:
Handing this to you, going to sleep.

Reply:
First, this all needs to be benchmarked again.
Second, I'm afraid we forgot about [nested embeds](https://github.com/prettier/prettier/issues/13158#issuecomment-1201057223). When the outer `embed` is called, the inner one should be already resolved. Both the inner and the outer one can be sync or async in any combination.

Reply:
```text
>node bin/prettier test.js
function HelloWorld() {
  return html`
    <h3>Bar List</h3>
    ${bars.map((bar) => html` <p>${bar}</p> `)}
  `;
}
```

Seems works

Reply:
Because they both are async in this case.

Reply:
I thought this not work

```js
html`
  <h3>Bar List</h3>
  ${bars.map(
    (bar) => html`
      <p
        style="width:        1px
"
      >
        a
      </p>
    `,
  )}
`;
```

https://github.com/prettier/prettier/blob/765cae25522f582b71a79e7fb10417eb4612abe6/src/language-html/embed.js#L90

Maybe it's time to remove that parentParser check.

Reply:
I didn't know (or forgot) about the check, but sorry, I don't see what's your point about it? Is it related to calling `embed` in the right order?

Reply:
I mean seems not broken.

Reply:
The outer one needs to be sync and the inner one async for the bug to manifest itself. Probably this situation currently never happens with the builtin printers, but it would still be a bug.

Reply:
As we're breakingly changing the signature of `embed` anyway (the textToDoc parameter becomes async), I think we should simply forbid sync embeds, i.e. the return type should become `undefined | () => Promise`.

Reply:
Corrently use sync embed in a few places.

https://github.com/fisker/prettier/blob/03db215c07c3013c0bc598c0cf7c57a7cd5c8230/src/language-html/embed.js#L379

Reply:
They can be sync, but they shouldn't use `print` in that case (because it's `print` that can try to get things from `embeds` that haven't become available yet). Should we probably remove `print` from the parameters of `embed` and pass it only to the returned functions then?

Reply:
> they shouldn't use print in that case

When calling `print` in the returned function, it's still possible not settled?


> Should we probably remove print

In that case, why not pass both `print` and `textToDoc` to the returned function, to prevent 

```js
const promise = textToDoc();
return () => promise
```

Reply:
> When calling print in the returned function, it's still possible not settled? Visting siblings doc, for example

I guess we should just say that it's a design limitation: `print` should be used only to print the current node's descendentans.

> In that case, why not pass both `print` and `textToDoc`

👍